packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -272,245 +340,310 @@ export const GridRootStyles = styled('div', {
+    [`& .${c['columnHeader--dragging']}, & .${c['row--dragging']}`]: {
+      background: (t.vars || t).palette.background.paper,
       padding: '0 12px',
       borderRadius: 'var(--unstable_DataGrid-radius)',
-      opacity: (theme.vars || theme).palette.action.disabledOpacity,
+      opacity: (t.vars || t).palette.action.disabledOpacity,
     },
-    [`& .${gridClasses['row--dragging']}`]: {
-      background: (theme.vars || theme).palette.background.paper,
+    [`& .${c['row--dragging']}`]: {
+      background: (t.vars || t).palette.background.paper,
       padding: '0 12px',
       borderRadius: 'var(--unstable_DataGrid-radius)',
-      opacity: (theme.vars || theme).palette.action.disabledOpacity,
+      opacity: (t.vars || t).palette.action.disabledOpacity,
 
-      [`& .${gridClasses.rowReorderCellPlaceholder}`]: {
+      [`& .${c.rowReorderCellPlaceholder}`]: {
         display: 'flex',
       },
     },
-    [`& .${gridClasses.treeDataGroupingCell}`]: {
+    [`& .${c['row--lastVisible']}`]: {
+      minHeight: 'unset !important',
+      maxHeight: 'unset !important',
+    },
+    [`& .${c['row--lastVisible']} .${c.cell}`]: {
+      width: 'var(--width)',
+      height: 'calc(var(--height) + var(--DataGrid-hasScrollX) * var(--DataGrid-scrollbarSize))',
+      paddingBottom: 'calc(var(--DataGrid-hasScrollX) * var(--DataGrid-scrollbarSize))',
+    },
+    [`& .${c.treeDataGroupingCell}`]: {
       display: 'flex',
       alignItems: 'center',
       width: '100%',
     },
-    [`& .${gridClasses.treeDataGroupingCellToggle}`]: {
+    [`& .${c.treeDataGroupingCellToggle}`]: {
       flex: '0 0 28px',
       alignSelf: 'stretch',
-      marginRight: theme.spacing(2),
+      marginRight: t.spacing(2),
     },
-    [`& .${gridClasses.groupingCriteriaCell}`]: {
+    [`& .${c.groupingCriteriaCell}`]: {
       display: 'flex',
       alignItems: 'center',
       width: '100%',
     },
-    [`& .${gridClasses.groupingCriteriaCellToggle}`]: {
+    [`& .${c.groupingCriteriaCellToggle}`]: {
       flex: '0 0 28px',
       alignSelf: 'stretch',
-      marginRight: theme.spacing(2),
+      marginRight: t.spacing(2),
     },
   };
 
   return gridStyle;
 });
+
+/**
+ * Merge a transparent overlay color with a background color, resulting in a single
+ * RGB color. The color space is gamma-corrected with a standard value of 2.2.
+ */
+function merge(background: string, overlay: string, opacity: number) {
+  const GAMMA = 2.2;
+
+  const f = (b: number, o: number) =>
+    ~~((b ** (1 / GAMMA) * (1 - opacity) + o ** (1 / GAMMA) * opacity) ** GAMMA);
+
+  const backgroundColor = decomposeColor(background);
+  const overlayColor = decomposeColor(overlay);
+
+  const rgb = [
+    f(backgroundColor.values[0], overlayColor.values[0]),
+    f(backgroundColor.values[1], overlayColor.values[1]),
+    f(backgroundColor.values[2], overlayColor.values[2]),
+  ] as const;
+
+  return recomposeColor({
+    type: 'rgb',
+    values: rgb as any,
+  });
+}
```

Comment:
This should probably go in the core's `colorManipulator.js`, but it would delay this PR to wait for it so I'll move it there after we merge this.